### ISSUE TITLE
Create good input vs bad input benchmarks?

### ISSUE THREAD

**Issue by azureskydiver** [azureskydiver/MatchChallenge#3]
It may be interesting to see how the various algorithms react to good inputs vs. bad inputs instead of the current approach of running a single benchmark passing in one large good input and and equally sized bad input. It may also be interesting to see how they handle different sized inputs.

**Comment by JohnHarald** [comment]
I was thinking the same when I made the comment in last pull request. About bimodal, I'm guessing it means KMP performs equally well for bad and good input (two balanced modes?), while the others mainly have their strength for good input (when there's a match). Attached a sample showing this.
[stat good bad.txt](https://github.com/azureskydiver/MatchChallenge/files/7777889/stat.good.bad.txt)
More tests takes lots more time though...

**Comment by azureskydiver** [comment]
I started a branch playing with BenchmarkDotNet's `Params` attribute, but I think I have too many possible values to be useful. Feel free to play with it and maybe tweak the parameters that maybe interesting.

**Comment by JohnHarald** [comment]
Yeah, a test is ran for each combination of params for each method. Just two options each of those three means 8 tests per method, a 5 minutes run quickly becomes an hour one. 
I'd say maybe include GoodInput bool, and a single default for BaseSize and Repetion?

The downside to GoodInput bool is that mixed input is not compared. It is basically the sum of good/bad, but you don't get to see that number in stat.

**Comment by JohnHarald** [comment]
We could apply ShortRunJob for shorter testing (twice speed).

Input type could be enum, either good/bad/mixed (preferred, easier to run all three) or good/bad as flags perhaps.

**Comment by azureskydiver** [comment]
Excellent idea!

**Comment by JohnHarald** [comment]
I have an update using enum that I could push to 'parameterized' branch.

**Comment by azureskydiver** [comment]
Yeah sorry for the misspelling. As I said yesterday, I was not thinking too clearly.

Go ahead and push, or start a new branch if you feel it would look cleaner

**Comment by JohnHarald** [comment]
I added to 'parameterized' branch.

Also added BenchmarkCategory for easier filtering which methods to benchmark using AnyCategoriesFilter/AllCategoriesFilter. I grouped the methods by letters and classified them with numbers (1 is best variant, 2 is other variants). It is then easy to for example only test "1" the best variants in each group, or for example only groups "A" and "E". This is especially useful when adding more parameter values which increases number of tests and one want to focus on some methods. (I have been commenting methods in and out during tests and that is tiresome)
